# 02 — Reconstruct Execution Episodes

This notebook converts window-level observations into execution-level records.

A raw observation is one machine/time-window record. Multiple observations may belong to the same reconstructed execution. The final `execution_id` will be created here, not inherited directly from `collection_id + instance_index`.

In [1]:
from pathlib import Path

import numpy as np
import pandas as pd


# ---------------------------------------------------------
# Locate repository root
# ---------------------------------------------------------

try:
    current_directory = Path.cwd()
except FileNotFoundError:
    current_directory = Path(
        "/Users/meghanakoti/masterproject/masters-project"
    )

REPO_ROOT = current_directory

while (
    REPO_ROOT.name != "masters-project"
    and REPO_ROOT.parent != REPO_ROOT
):
    REPO_ROOT = REPO_ROOT.parent

if REPO_ROOT.name != "masters-project":
    fallback_root = Path(
        "/Users/meghanakoti/masterproject/masters-project"
    )

    if fallback_root.exists():
        REPO_ROOT = fallback_root
    else:
        raise RuntimeError(
            "Could not locate repository root named "
            "'masters-project'."
        )


# ---------------------------------------------------------
# Locate canonical observation-level dataset
# ---------------------------------------------------------

OBSERVATIONS_PATH = (
    REPO_ROOT
    / "data"
    / "processed"
    / "google_subset_v1"
    / "google_subset_observations_v1.parquet"
)

if not OBSERVATIONS_PATH.exists():
    processed_directory = OBSERVATIONS_PATH.parent

    available_files = (
        sorted(processed_directory.iterdir())
        if processed_directory.exists()
        else []
    )

    raise FileNotFoundError(
        "Canonical observations file was not found:\n"
        f"{OBSERVATIONS_PATH}\n\n"
        "Available files in the processed directory:\n"
        + (
            "\n".join(
                str(file)
                for file in available_files
            )
            if available_files
            else "The processed directory does not exist."
        )
        + "\n\n"
        "Run the final save cell in Notebook 01 first."
    )


# ---------------------------------------------------------
# Load observation-level dataset
# ---------------------------------------------------------

df = pd.read_parquet(OBSERVATIONS_PATH)

pd.set_option("display.max_columns", None)
pd.set_option("display.width", 160)

print("Repository root:", REPO_ROOT)
print("Loaded:", OBSERVATIONS_PATH)
print("Shape:", df.shape)

print("\nColumns:")
print(df.columns.tolist())

print("\nSample rows:")
display(df.head())

Repository root: c:\Users\012320802\masters-project
Loaded: c:\Users\012320802\masters-project\data\processed\google_subset_v1\google_subset_observations_v1.parquet
Shape: (405894, 51)

Columns:
['Unnamed: 0', 'time', 'instance_events_type', 'collection_id', 'scheduling_class', 'collection_type', 'priority', 'alloc_collection_id', 'instance_index', 'machine_id', 'resource_request', 'constraint', 'collections_events_type', 'user', 'collection_name', 'collection_logical_name', 'start_after_collection_ids', 'vertical_scaling', 'scheduler', 'start_time', 'end_time', 'average_usage', 'maximum_usage', 'random_sample_usage', 'assigned_memory', 'page_cache_memory', 'cycles_per_instruction', 'memory_accesses_per_instruction', 'sample_rate', 'cpu_usage_distribution', 'tail_cpu_usage_distribution', 'cluster', 'event', 'failed', 'raw_req_cpu', 'raw_req_memory', 'raw_avg_cpu', 'raw_avg_memory', 'raw_peak_cpu', 'raw_peak_memory', 'event_time_us', 'usage_start_time_us', 'usage_end_time_us', 'usage_wi

,Unnamed: 0,time,instance_events_type,collection_id,scheduling_class,collection_type,priority,alloc_collection_id,instance_index,machine_id,resource_request,constraint,collections_events_type,user,collection_name,collection_logical_name,start_after_collection_ids,vertical_scaling,scheduler,start_time,end_time,average_usage,maximum_usage,random_sample_usage,assigned_memory,page_cache_memory,cycles_per_instruction,memory_accesses_per_instruction,sample_rate,cpu_usage_distribution,tail_cpu_usage_distribution,cluster,event,failed,raw_req_cpu,raw_req_memory,raw_avg_cpu,raw_avg_memory,raw_peak_cpu,raw_peak_memory,event_time_us,usage_start_time_us,usage_end_time_us,usage_window_duration_sec,requested_cpu_ncu,requested_memory_normalized,average_cpu_ncu,average_memory_normalized,peak_cpu_ncu,peak_memory_normalized,observation_id
0,0,0,2,94591244395,3,1,200,0,144,168846390496,"{'cpus': 0.020660400390625, 'memory': 0.014434...",[],2,fn8Ve4Tdl/FVVvwXFGIKe4+Wo4zLjUL/557qdFVYu5M=,Hzsv/gF8CPQXdqpsfovDTC1TJNyphDxPu7vaTeNxA74=,YCuhYrnORLiUh9WGL5q5tkBevfwtucSnFr2qPZh6Kes=,[],1.0,0.0,274800000000,275100000000,"{'cpus': 0.00466156005859375, 'memory': 0.0059...","{'cpus': 0.01190185546875, 'memory': 0.0059356...","{'cpus': 0.0043487548828125, 'memory': None}",0.014435,0.000415,NaN,NaN,1.0,[0.00314331 0.00381088 0.00401306 0.00415039 0...,[0.00535583 0.00541687 0.00548553 0.00554657 0...,7,FAIL,1,0.020660,0.014435,0.004662,5.920410e-03,0.011902,5.935669e-03,0,274800000000,275100000000,300.0,0.020660,0.014435,0.004662,5.920410e-03,0.011902,5.935669e-03,94591244395||144||168846390496||274800000000||...
1,1,2517305308183,2,260697606809,2,0,360,221495397286,335,85515092,"{'cpus': 0.00724029541015625, 'memory': 0.0013...",[],2,DrrEIEWkWuW7RrZwpHLCN0k0A2J0usJeyt3wtqzZ7Kk=,hDGffcrF/rhQQEG8Uns/RMUK7R15DXjFnRasoKFhefI=,wcRcAMuop2OqH9EW4feH919tadFec5a11ply0hcS/C8=,[],2.0,0.0,1800713000000,1800714000000,"{'cpus': 0.0, 'memory': 9.5367431640625e-07}","{'cpus': 0.0, 'memory': 9.5367431640625e-07}","{'cpus': 0.0, 'memory': None}",0.000000,0.000000,NaN,NaN,1.0,[1.23977661e-05 1.23977661e-05 1.23977661e-05 ...,[1.23977661e-05 1.23977661e-05 1.23977661e-05 ...,7,FAIL,1,0.007240,0.001303,0.000000,9.536743e-07,0.000000,9.536743e-07,2517305308183,1800713000000,1800714000000,1.0,0.007240,0.001303,0.000000,9.536743e-07,0.000000,9.536743e-07,260697606809||335||85515092||1800713000000||18...
2,2,195684022913,6,276227177776,2,0,103,0,376,169321752432,"{'cpus': 0.048583984375, 'memory': 0.004165649...",[],6,/ivQBmewiFcXfGJdCUsEKx47NiRE29Tjiq3gw+zR2Cg=,kk6+maA6fvAdJ+VTU8AcpzQPTyVrx+ySt0MXRAyO8FU=,zCA2dl2PDptd82Hob906gE82JHzx0SbqA4mZurqZdmY=,[],2.0,1.0,81300000000,81600000000,"{'cpus': 0.024200439453125, 'memory': 0.002788...","{'cpus': 0.06005859375, 'memory': 0.0028457641...","{'cpus': 0.026458740234375, 'memory': None}",0.010422,0.000235,0.939919,0.001318,1.0,[0.01344299 0.01809692 0.0201416 0.02246094 0...,[0.02902222 0.02929688 0.0295105 0.0296936 0...,7,SCHEDULE,0,0.048584,0.004166,0.024200,2.788544e-03,0.060059,2.845764e-03,195684022913,81300000000,81600000000,300.0,0.048584,0.004166,0.024200,2.788544e-03,0.060059,2.845764e-03,276227177776||376||169321752432||81300000000||...
3,3,0,2,10507389885,3,0,200,0,1977,178294817221,"{'cpus': 0.0704345703125, 'memory': 0.04162597...",[],2,8qRmTJas/6XEBaA0l4Wt1+/qSLgc6p7u7JzoMSuT/M8=,fypwFjdqaQPSxCfeqVPCBAvFcnntmkRpQxwQ/vJsCxU=,uNjMQD1+DL9IgCFckx8lHOsCbyvLgKmZCmRjiyWZNhk=,[],2.0,0.0,1075500000000,1075800000000,"{'cpus': 0.047607421875, 'memory': 0.034423828...","{'cpus': 0.13330078125, 'memory': 0.03466796875}","{'cpus': 0.05084228515625, 'memory': None}",0.041626,0.000225,1.359102,0.007643,1.0,[0.03704834 0.04125977 0.04290771 0.04425049 0...,[0.05535889 0.05584717 0.05633545 0.05718994 0...,8,FAIL,1,0.070435,0.041626,0.047607,3.442383e-02,0.133301,3.466797e-02,0,1075500000000,1075800000000,300.0,0.070435,0.041626,0.047607,3.442383e-02,0.133301,3.466797e-02,10507389885||1977||178294817221||1075500000000...
4,4,1810627494172,3,25

## Step 2 — Validate Reconstruction Inputs

Before reconstructing executions, verify that every observation has the identifiers and timestamp fields required for grouping.

`observation_id` uniquely identifies one raw machine/time-window observation. It is not the final execution identifier.

In [2]:
#Validate reconstruction input
required_columns = [
    "observation_id",
    "collection_id",
    "instance_index",
    "machine_id",
    "event_time_us",
    "usage_start_time_us",
    "usage_end_time_us",
]

missing_columns = [
    column for column in required_columns
    if column not in df.columns
]

if missing_columns:
    raise ValueError(
        "The canonical observation dataset is missing:\n"
        + "\n".join(f"- {column}" for column in missing_columns)
    )

print("All required reconstruction columns are present.")
print("Total observations:", len(df))
print("Unique observation IDs:", df["observation_id"].nunique())
print(
    "Duplicate observation IDs:",
    df["observation_id"].duplicated().sum()
)

All required reconstruction columns are present.
Total observations: 405894
Unique observation IDs: 405894
Duplicate observation IDs: 0


## Step 3 — Prepare Canonical Timestamp Fields

Convert the canonical timestamp columns to numeric microsecond values and validate the usage intervals.

The original timestamp columns remain preserved. The derived `usage_duration_us` field represents the duration of an observation window, not necessarily the complete execution runtime.

In [3]:
timestamp_columns = [
    "event_time_us",
    "usage_start_time_us",
    "usage_end_time_us",
]

for column in timestamp_columns:
    df[column] = pd.to_numeric(df[column], errors="coerce")

invalid_timestamp_counts = df[timestamp_columns].isna().sum()

print("Missing canonical timestamps:")
display(invalid_timestamp_counts)

# Keep only rows with a valid usage interval
df = df.dropna(
    subset=["usage_start_time_us", "usage_end_time_us"]
).copy()

# Ensure interval ordering
df["usage_duration_us"] = (
    df["usage_end_time_us"] -
    df["usage_start_time_us"]
)

negative_duration_count = (
    df["usage_duration_us"] < 0
).sum()

print("Negative usage durations:", negative_duration_count)

if negative_duration_count > 0:
    raise ValueError(
        "Negative usage intervals detected. "
        "Inspect timestamps before reconstruction."
    )

df = df.sort_values(
    [
        "collection_id",
        "instance_index",
        "usage_start_time_us",
        "usage_end_time_us",
        "machine_id",
    ]
).reset_index(drop=True)

print("Prepared observations:", len(df))
display(
    df[
        [
            "observation_id",
            "collection_id",
            "instance_index",
            "machine_id",
            "event_time_us",
            "usage_start_time_us",
            "usage_end_time_us",
            "usage_duration_us",
        ]
    ].head()
)

Missing canonical timestamps:


event_time_us          3
usage_start_time_us    0
usage_end_time_us      0
dtype: int64

Negative usage durations: 0
Prepared observations: 405894


,observation_id,collection_id,instance_index,machine_id,event_time_us,usage_start_time_us,usage_end_time_us,usage_duration_us
0,6824||0||466339793||2001900000000||2002200000000,6824,0,466339793,0,2001900000000,2002200000000,300000000
1,6824||9||466365904||2001900000000||2002200000000,6824,9,466365904,0,2001900000000,2002200000000,300000000
2,6824||16||527302722||2001900000000||2002200000000,6824,16,527302722,0,2001900000000,2002200000000,300000000
3,6824||32||527367127||2001900000000||2002200000000,6824,32,527367127,0,2001900000000,2002200000000,300000000
4,6824||34||527365830||2001900000000||2002200000000,6824,34,527365830,0,2001900000000,2002200000000,300000000


## Step 4 — Audit Candidate Instance Groups

Use `(collection_id, instance_index)` as a candidate grouping key for inspection only.

This audit checks how many observations and machines appear in each candidate group and how wide its time span is. The results will determine whether additional episode separation is required before creating the final `execution_id`.

In [4]:
# This is only an audit key.
# It is not yet the final execution_id.
df["candidate_instance_key"] = (
    df["collection_id"].astype(str)
    + "||"
    + df["instance_index"].astype(str)
)

instance_audit = (
    df.groupby("candidate_instance_key")
      .agg(
          observation_count=("observation_id", "nunique"),
          machine_count=("machine_id", "nunique"),
          first_usage_start_us=("usage_start_time_us", "min"),
          last_usage_end_us=("usage_end_time_us", "max"),
      )
      .reset_index()
)

instance_audit["span_us"] = (
    instance_audit["last_usage_end_us"]
    - instance_audit["first_usage_start_us"]
)

print(
    "Unique candidate collection-instance groups:",
    len(instance_audit)
)

print("\nObservations per candidate group:")
display(instance_audit["observation_count"].describe())

print("\nMachines per candidate group:")
display(instance_audit["machine_count"].describe())

print("\nCandidate groups with multiple machines:")
display(
    instance_audit[
        instance_audit["machine_count"] > 1
    ].head(10)
)

print("\nLargest candidate groups:")
display(
    instance_audit.sort_values(
        "observation_count",
        ascending=False
    ).head(10)
)

Unique candidate collection-instance groups: 242946

Observations per candidate group:


count    242946.000000
mean          1.670717
std          17.788001
min           1.000000
25%           1.000000
50%           1.000000
75%           1.000000
max        7704.000000
Name: observation_count, dtype: float64


Machines per candidate group:


count    242946.000000
mean          1.620002
std          15.906318
min           1.000000
25%           1.000000
50%           1.000000
75%           1.000000
max        6640.000000
Name: machine_count, dtype: float64


Candidate groups with multiple machines:


,candidate_instance_key,observation_count,machine_count,first_usage_start_us,last_usage_end_us,span_us
103,10121672||1303,2,2,474900000000,475200000000,300000000
116,10121672||1470,2,2,474900000000,475200000000,300000000
163,10121672||2050,2,2,474900000000,475200000000,300000000
175,10121672||2220,2,2,474900000000,475200000000,300000000
196,10121672||2447,2,2,474900000000,475200000000,300000000
203,10121672||2516,2,2,474900000000,475200000000,300000000
237,10121672||738,2,2,474900000000,2031600000000,1556700000000
247,10121672||924,2,2,474900000000,2031600000000,1556700000000
559,10445557720||1850,2,2,375600000000,2670000000000,2294400000000
800,105142593258||1055,13,13,678000000000,678300000000,300000000



Largest candidate groups:


,candidate_instance_key,observation_count,machine_count,first_usage_start_us,last_usage_end_us,span_us
1927,116646838687||17,7704,6640,361200000000,2288923000000,1927723000000
7656,14423004927||969,1430,1430,1979700000000,1980000000000,300000000
2124,122697038507||1191,1308,1308,1944000000000,1944300000000,300000000
127632,348020181306||5,1135,1135,2544777000000,2544778000000,1000000
7580,14423004927||1021,994,994,1979700000000,1980000000000,300000000
49308,260725669813||176,905,903,25800000000,1175400000000,1149600000000
49234,260725669813||112,767,758,25800000000,1175400000000,1149600000000
2270,122697038507||5012,686,686,1944000000000,1944300000000,300000000
3770,128073966478||2432,587,586,115800000000,430500000000,314700000000
49241,260725669813||118,580,578,25800000000,1175400000000,1149600000000


In [5]:
# Sort observations chronologically within each candidate group
gap_df = df.sort_values(
    [
        "candidate_instance_key",
        "usage_start_time_us",
        "usage_end_time_us",
    ]
).copy()

# Previous observation end within the same candidate group
gap_df["previous_usage_end_us"] = (
    gap_df
    .groupby("candidate_instance_key")["usage_end_time_us"]
    .shift(1)
)

# Time gap between consecutive observations
gap_df["gap_from_previous_us"] = (
    gap_df["usage_start_time_us"]
    - gap_df["previous_usage_end_us"]
)

# Exclude the first observation in each group
positive_gaps = gap_df.loc[
    gap_df["gap_from_previous_us"].notna()
    & (gap_df["gap_from_previous_us"] > 0),
    "gap_from_previous_us",
]

print("Observations with a previous observation:", gap_df["previous_usage_end_us"].notna().sum())
print("Positive temporal gaps:", len(positive_gaps))

print("\nPositive gap distribution in microseconds:")
display(
    positive_gaps.describe(
        percentiles=[0.50, 0.75, 0.90, 0.95, 0.99, 0.999]
    )
)

print("\nLargest temporal gaps:")
display(
    gap_df[
        gap_df["gap_from_previous_us"].notna()
    ]
    .sort_values("gap_from_previous_us", ascending=False)
    [
        [
            "candidate_instance_key",
            "observation_id",
            "machine_id",
            "previous_usage_end_us",
            "usage_start_time_us",
            "gap_from_previous_us",
        ]
    ]
    .head(20)
)

Observations with a previous observation: 162948
Positive temporal gaps: 10639

Positive gap distribution in microseconds:


count                 10639.0
mean      599492682018.986694
std       585178365982.275879
min              1500000000.0
50%            346200000000.0
75%           1047300000000.0
90%           1521900000000.0
95%           1774013000000.0
99%      2222692340000.015625
99.9%         2400000000000.0
max           2530500000000.0
Name: gap_from_previous_us, dtype: Float64


Largest temporal gaps:


,candidate_instance_key,observation_id,machine_id,previous_usage_end_us,usage_start_time_us,gap_from_previous_us
57908,96110849886||4625,96110849886||4625||2272276147||2537100000000||...,2272276147,6600000000,2537100000000,2530500000000
83431,134216236214||474,134216236214||474||225350648716||2561700000000...,225350648716,138900000000,2561700000000,2422800000000
83433,134216236214||507,134216236214||507||52214052853||2561700000000|...,52214052853,138900000000,2561700000000,2422800000000
232529,330587274558||1269,330587274558||1269||20736934||2677800000000||2...,20736934,277800000000,2677800000000,2400000000000
232482,330587274558||145,330587274558||145||4646834074||2677800000000||...,4646834074,277800000000,2677800000000,2400000000000
232536,330587274558||1430,330587274558||1430||1644702886||2677800000000|...,1644702886,277800000000,2677800000000,2400000000000
232508,330587274558||678,330587274558||678||4451026018||2677800000000||...,4451026018,277800000000,2677800000000,2400000000000
232502,330587274558||629,330587274558||629||22672367||2677800000000||26...,22672367,277800000000,2677800000000,2400000000000
232478,330587274558||38,330587274558||38||23624179225||2677800000000||...,23624179225,277800000000,2677800000000,2400000000000
232523,330587274558||1177,330587274558||1177||21284286||2677800000000||2...,21284286,277800000000,2677800000000,2400000000000


## Step 6 — Interpret Temporal Gaps

Convert the temporal gaps from microseconds into minutes and days.

This helps distinguish normal spacing between usage windows from larger inactive periods that may indicate a new execution episode.

In [6]:
gap_summary = gap_df[
    gap_df["gap_from_previous_us"].notna()
    & (gap_df["gap_from_previous_us"] > 0)
].copy()

gap_summary["gap_minutes"] = (
    gap_summary["gap_from_previous_us"] / 60_000_000
)

gap_summary["gap_days"] = (
    gap_summary["gap_from_previous_us"] / 86_400_000_000
)

print("Positive gap summary in minutes:")
display(
    gap_summary["gap_minutes"].describe(
        percentiles=[0.50, 0.75, 0.90, 0.95, 0.99, 0.999]
    )
)

print("Positive gap summary in days:")
display(
    gap_summary["gap_days"].describe(
        percentiles=[0.50, 0.75, 0.90, 0.95, 0.99, 0.999]
    )
)

print("Most common positive gaps:")
display(
    gap_summary["gap_minutes"]
    .round(3)
    .value_counts()
    .head(20)
)

Positive gap summary in minutes:


count         10639.0
mean        9991.5447
std       9752.972766
min              25.0
50%            5770.0
75%           17455.0
90%           25365.0
95%      29566.883333
99%      37044.872333
99.9%         40000.0
max           42175.0
Name: gap_minutes, dtype: Float64

Positive gap summary in days:


count      10639.0
mean      6.938573
std       6.772898
min       0.017361
50%       4.006944
75%      12.121528
90%      17.614583
95%      20.532558
99%      25.725606
99.9%    27.777778
max      29.288194
Name: gap_days, dtype: Float64

Most common positive gaps:


gap_minutes
2420.0      1528
19540.0      598
1660.0       441
18965.0      276
25365.0      215
4206.85      170
15720.0      167
1790.0       156
1675.567     121
940.0        117
17455.0      117
15681.95     113
3256.683     112
20335.0      103
33.083       102
10245.0      102
1540.0        90
6185.0        88
545.0         88
9962.033      85
Name: count, dtype: Int64

Above:
The smallest positive gap is 25 minutes.
The median positive gap is approximately 4 days.
Most positive gaps represent long inactive periods, not normal adjacent observation windows.
Therefore, we can use a 5-minute continuity threshold, matching the approximate observation-window duration.

Any gap greater than 5 minutes will start a new execution episode.

## Step 7 — Define Execution Episode Boundaries

The usage observations represent approximately five-minute windows.

The smallest positive gap between observations is 25 minutes, which is already larger than one normal observation window. Therefore, a gap greater than five minutes is treated as a boundary between execution episodes.

A machine change alone does not create a new episode. The episode boundary is based on a temporal gap within the same candidate collection-instance group.

In [7]:
# Approximate observation-window duration
OBSERVATION_WINDOW_US = 5 * 60 * 1_000_000  # five minutes

# Validate the observed positive gaps
minimum_positive_gap_us = (
    gap_df.loc[
        gap_df["gap_from_previous_us"].notna()
        & (gap_df["gap_from_previous_us"] > 0),
        "gap_from_previous_us"
    ].min()
)

print(
    "Minimum positive gap in minutes:",
    minimum_positive_gap_us / 60_000_000
)

if minimum_positive_gap_us <= OBSERVATION_WINDOW_US:
    print(
        "Warning: some positive gaps are within one observation window."
    )
else:
    print(
        "All positive gaps exceed the five-minute observation window."
    )

# A new episode begins after a gap larger than one observation window.
gap_df["new_episode"] = (
    gap_df["gap_from_previous_us"] > OBSERVATION_WINDOW_US
)

# The first observation in each candidate group starts episode zero.
gap_df["new_episode"] = (
    gap_df["new_episode"]
    | gap_df["previous_usage_end_us"].isna()
)

# Number episodes chronologically within each candidate group.
gap_df["episode_number"] = (
    gap_df
    .groupby("candidate_instance_key")["new_episode"]
    .cumsum()
    .astype(int)
)

# Create the final reconstructed execution ID.
gap_df["execution_id"] = (
    gap_df["candidate_instance_key"]
    + "||episode_"
    + gap_df["episode_number"].astype(str)
)

print("Reconstructed execution episodes:")
print("Unique execution IDs:", gap_df["execution_id"].nunique())
print("Total observations:", len(gap_df))

Minimum positive gap in minutes: 25.0
All positive gaps exceed the five-minute observation window.
Reconstructed execution episodes:
Unique execution IDs: 253585
Total observations: 405894


## Step 8 — Create the Reconstructed Execution ID

The dataset does not provide an explicit episode ID.

In the previous step, we inferred episodes by:

- grouping observations by `collection_id + instance_index`;
- ordering them by time;
- starting a new episode when the time gap was greater than five minutes.

This cell creates the final `execution_id` by combining the candidate group with its episode number.

For example:

```text
candidate group: 100||7
episode 0:       100||7||episode_0
episode 1:       100||7||episode_1

In [8]:
# Reload the canonical observation dataset to remove
# any partially merged columns from the previous attempt.
base_df = pd.read_parquet(OBSERVATIONS_PATH)

# Remove older or previously generated reconstruction columns.
base_df = base_df.drop(
    columns=[
        "execution_id",
        "candidate_instance_key",
        "episode_number",
    ],
    errors="ignore",
)

# Create a clean lookup table from the temporal reconstruction.
execution_lookup = (
    gap_df[
        [
            "observation_id",
            "candidate_instance_key",
            "episode_number",
            "execution_id",
        ]
    ]
    .drop_duplicates("observation_id")
)

# Join reconstructed execution information back to each observation.
df = base_df.merge(
    execution_lookup,
    on="observation_id",
    how="left",
    validate="one_to_one",
)

# Validate the assignment.
print("Execution ID assignment complete.")
print("Total observations:", len(df))
print(
    "Missing reconstructed execution IDs:",
    df["execution_id"].isna().sum()
)
print(
    "Unique reconstructed executions:",
    df["execution_id"].nunique()
)

display(
    df[
        [
            "observation_id",
            "candidate_instance_key",
            "episode_number",
            "execution_id",
        ]
    ].head()
)

Execution ID assignment complete.
Total observations: 405894
Missing reconstructed execution IDs: 0
Unique reconstructed executions: 253585


,observation_id,candidate_instance_key,episode_number,execution_id
0,94591244395||144||168846390496||274800000000||...,94591244395||144,1,94591244395||144||episode_1
1,260697606809||335||85515092||1800713000000||18...,260697606809||335,1,260697606809||335||episode_1
2,276227177776||376||169321752432||81300000000||...,276227177776||376,1,276227177776||376||episode_1
3,10507389885||1977||178294817221||1075500000000...,10507389885||1977,1,10507389885||1977||episode_1
4,25911621841||3907||231364893292||1565315000000...,25911621841||3907,1,25911621841||3907||episode_1


## Step 9 — Aggregate Observations Into Reconstructed Execution Episodes

This step creates one record per reconstructed execution episode.
- Requested CPU and memory are counted once per machine and summed across the execution.
- Actual CPU and memory usage are summed across machines within each time window.
- The maximum summed usage across time windows is used as the execution-level peak.
- Missing resource measurements are preserved as unknown rather than treated as zero.
- Execution-level records are validated for uniqueness, runtime consistency, missing resource values, and multi-machine executions.

In [9]:
# =========================================================
# Step 9: Create and Validate Execution-Level Records
# =========================================================

# Required columns for execution-level aggregation
required_columns = [
    "execution_id",
    "collection_id",
    "instance_index",
    "candidate_instance_key",
    "episode_number",
    "observation_id",
    "machine_id",
    "usage_start_time_us",
    "usage_end_time_us",
    "requested_cpu_ncu",
    "requested_memory_normalized",
    "average_cpu_ncu",
    "average_memory_normalized",
    "peak_cpu_ncu",
    "peak_memory_normalized",
]

missing_columns = [
    column for column in required_columns
    if column not in df.columns
]

if missing_columns:
    raise ValueError(
        f"Missing required columns: {missing_columns}"
    )


# =========================================================
# Aggregate requested resources
# =========================================================

# Count each machine's request once within an execution.
machine_requests = (
    df.groupby(
        ["execution_id", "machine_id"],
        as_index=False,
        dropna=False,
    )
    .agg(
        requested_cpu_ncu=(
            "requested_cpu_ncu",
            "first",
        ),
        requested_memory_normalized=(
            "requested_memory_normalized",
            "first",
        ),
    )
)

# Sum machine requests across the execution.
# min_count=1 preserves all-missing values as NaN
# instead of treating them as zero.
execution_requests = (
    machine_requests
    .groupby("execution_id", as_index=False)
    .agg(
        requested_cpu_ncu=(
            "requested_cpu_ncu",
            lambda x: x.sum(min_count=1),
        ),
        requested_memory_normalized=(
            "requested_memory_normalized",
            lambda x: x.sum(min_count=1),
        ),
    )
)


# =========================================================
# Aggregate actual CPU and memory usage
# =========================================================

# Sum actual usage across machines within each time window.
# min_count=1 preserves an all-missing window as NaN.
window_usage = (
    df.groupby(
        [
            "execution_id",
            "usage_start_time_us",
            "usage_end_time_us",
        ],
        as_index=False,
        dropna=False,
    )
    .agg(
        total_cpu_ncu=(
            "peak_cpu_ncu",
            lambda x: x.sum(min_count=1),
        ),
        total_memory_normalized=(
            "peak_memory_normalized",
            lambda x: x.sum(min_count=1),
        ),
    )
)

# The execution peak is the maximum total usage
# observed across all time windows.
execution_peaks = (
    window_usage
    .groupby("execution_id", as_index=False)
    .agg(
        peak_cpu_ncu=(
            "total_cpu_ncu",
            "max",
        ),
        peak_memory_normalized=(
            "total_memory_normalized",
            "max",
        ),
    )
)


# =========================================================
# Create one row per reconstructed execution episode
# =========================================================

execution_df = (
    df.groupby("execution_id", as_index=False)
      .agg(
          collection_id=("collection_id", "first"),
          instance_index=("instance_index", "first"),
          candidate_instance_key=(
              "candidate_instance_key",
              "first",
          ),
          episode_number=("episode_number", "first"),

          execution_start_time_us=(
              "usage_start_time_us",
              "min",
          ),
          execution_end_time_us=(
              "usage_end_time_us",
              "max",
          ),

          observation_count=(
              "observation_id",
              "nunique",
          ),
          machine_count=(
              "machine_id",
              "nunique",
          ),

          average_cpu_ncu=(
              "average_cpu_ncu",
              "mean",
          ),
          average_memory_normalized=(
              "average_memory_normalized",
              "mean",
          ),
      )
)


# =========================================================
# Attach corrected execution-level resource values
# =========================================================

execution_df = execution_df.merge(
    execution_requests,
    on="execution_id",
    how="left",
    validate="one_to_one",
)

execution_df = execution_df.merge(
    execution_peaks,
    on="execution_id",
    how="left",
    validate="one_to_one",
)


# =========================================================
# Runtime and resource slack
# =========================================================

execution_df["runtime_us"] = (
    execution_df["execution_end_time_us"]
    - execution_df["execution_start_time_us"]
)

execution_df["runtime_sec"] = (
    execution_df["runtime_us"] / 1_000_000
)

execution_df["slack_cpu_ncu"] = (
    execution_df["requested_cpu_ncu"]
    - execution_df["peak_cpu_ncu"]
)

execution_df["slack_memory_normalized"] = (
    execution_df["requested_memory_normalized"]
    - execution_df["peak_memory_normalized"]
)


# =========================================================
# Validate all reconstructed execution episodes
# =========================================================

print("Execution-level dataset created.")
print("Total execution records:", len(execution_df))
print(
    "Unique execution IDs:",
    execution_df["execution_id"].nunique()
)
print(
    "Duplicate execution IDs:",
    execution_df["execution_id"].duplicated().sum()
)
print(
    "Missing execution IDs:",
    execution_df["execution_id"].isna().sum()
)
print(
    "Negative runtimes:",
    (execution_df["runtime_us"] < 0).sum()
)
print(
    "Missing runtimes:",
    execution_df["runtime_sec"].isna().sum()
)

# Resource aggregation validation
print("\nResource aggregation validation:")
print(
    "Missing requested CPU:",
    execution_df["requested_cpu_ncu"].isna().sum()
)
print(
    "Missing requested memory:",
    execution_df["requested_memory_normalized"].isna().sum()
)
print(
    "Missing peak CPU:",
    execution_df["peak_cpu_ncu"].isna().sum()
)
print(
    "Missing peak memory:",
    execution_df["peak_memory_normalized"].isna().sum()
)
print(
    "Multi-machine executions:",
    (execution_df["machine_count"] > 1).sum()
)


# Show how many records exist for each episode number.
print("\nExecution records by episode number:")

display(
    execution_df["episode_number"]
    .value_counts()
    .sort_index()
    .rename_axis("episode_number")
    .reset_index(name="execution_count")
)


# Find candidate groups containing multiple episodes.
episodes_per_candidate = (
    execution_df
    .groupby("candidate_instance_key")["episode_number"]
    .nunique()
    .reset_index(name="episode_count")
)

multi_episode_candidates = (
    episodes_per_candidate[
        episodes_per_candidate["episode_count"] > 1
    ]
    .sort_values("episode_count", ascending=False)
)

print(
    "\nCandidate groups with multiple episodes:",
    len(multi_episode_candidates)
)

display(multi_episode_candidates.head(20))


# Display execution records after episode 1.
print("\nExamples of episode_2 or later:")

display(
    execution_df[
        execution_df["episode_number"] > 1
    ]
    .sort_values(
        ["candidate_instance_key", "episode_number"]
    )
    [
        [
            "execution_id",
            "candidate_instance_key",
            "episode_number",
            "execution_start_time_us",
            "execution_end_time_us",
            "runtime_sec",
            "observation_count",
            "machine_count",
        ]
    ]
    .head(20)
)


# Display the largest reconstructed execution episodes.
print("\nLargest reconstructed execution episodes:")

display(
    execution_df
    .sort_values(
        ["observation_count", "machine_count"],
        ascending=False,
    )
    [
        [
            "execution_id",
            "candidate_instance_key",
            "episode_number",
            "runtime_sec",
            "observation_count",
            "machine_count",
            "requested_cpu_ncu",
            "requested_memory_normalized",
            "peak_cpu_ncu",
            "peak_memory_normalized",
        ]
    ]
    .head(20)
)

Execution-level dataset created.
Total execution records: 253585
Unique execution IDs: 253585
Duplicate execution IDs: 0
Missing execution IDs: 0
Negative runtimes: 0
Missing runtimes: 0

Resource aggregation validation:
Missing requested CPU: 774
Missing requested memory: 774
Missing peak CPU: 0
Missing peak memory: 4495
Multi-machine executions: 38143

Execution records by episode number:


,episode_number,execution_count
0,1,242946
1,2,10291
2,3,340
3,4,8



Candidate groups with multiple episodes: 10291


,candidate_instance_key,episode_count
49175,260697606809||667,4
30899,221495509609||651,4
48950,260697606809||1114,4
49177,260697606809||676,4
49187,260697606809||749,4
49060,260697606809||2032,4
49208,260697606809||933,4
49018,260697606809||1633,4
30925,221495509609||901,3
30778,221495509609||1875,3



Examples of episode_2 or later:


,execution_id,candidate_instance_key,episode_number,execution_start_time_us,execution_end_time_us,runtime_sec,observation_count,machine_count
102,10121672||1300||episode_2,10121672||1300,2,2031300000000,2031600000000,300.0,1,1
105,10121672||1307||episode_2,10121672||1307,2,2031300000000,2031600000000,300.0,1,1
240,10121672||738||episode_2,10121672||738,2,2031300000000,2031600000000,300.0,1,1
251,10121672||924||episode_2,10121672||924,2,2031300000000,2031600000000,300.0,1,1
545,10445557720||162||episode_2,10445557720||162,2,2669700000000,2670000000000,300.0,1,1
565,10445557720||1850||episode_2,10445557720||1850,2,2669700000000,2670000000000,300.0,1,1
606,10445557720||2394||episode_2,10445557720||2394,2,2669700000000,2670000000000,300.0,1,1
643,10445557720||684||episode_2,10445557720||684,2,2669700000000,2670000000000,300.0,1,1
1121,1123179998||390||episode_2,1123179998||390,2,2158500000000,2158800000000,300.0,1,1
1227,114466593044||118||episode_2,114466593044||118,2,1627200000000,1627500000000,300.0,1,1



Largest reconstructed execution episodes:


,execution_id,candidate_instance_key,episode_number,runtime_sec,observation_count,machine_count,requested_cpu_ncu,requested_memory_normalized,peak_cpu_ncu,peak_memory_normalized
1947,116646838687||17||episode_2,116646838687||17,2,1.0,6566,6566,7.936203,42.578171,0.000000,0.006262
8134,14423004927||969||episode_1,14423004927||969,1,300.0,1430,1430,1.712280,0.283382,0.924625,0.147285
2155,122697038507||1191||episode_1,122697038507||1191,1,300.0,1308,1308,17.284058,6.815826,32.731934,4.540558
1946,116646838687||17||episode_1,116646838687||17,1,17.0,1138,1138,1.293655,7.405960,6.320679,0.863884
134845,348020181306||5||episode_1,348020181306||5,1,1.0,1135,1135,0.000000,14.772873,0.000000,0.001082
8056,14423004927||1021||episode_1,14423004927||1021,1,300.0,994,994,1.187342,0.196586,0.642712,0.102379
53392,260725669813||176||episode_1,260725669813||176,1,300.0,903,903,9.779491,1.176355,1.257305,4.781204
53271,260725669813||112||episode_1,260725669813||112,1,300.0,758,758,8.200031,0.987461,1.055412,4.013458
2303,122697038507||5012||episode_1,122697038507||5012,1,300.0,686,686,9.064163,3.573952,17.166748,2.381363
3896,128073966478||2432||episode_1,128073966478||2432,1,241.0,586,586,1.477562,0.130003,2.847916,0.107300


In [10]:
# =========================================================
# Validate missing-value handling for resource aggregation
# =========================================================

print("Source resource missing values:")
print("requested CPU:", df["requested_cpu_ncu"].isna().sum())
print("requested memory:", df["requested_memory_normalized"].isna().sum())
print("peak CPU:", df["peak_cpu_ncu"].isna().sum())
print("peak memory:", df["peak_memory_normalized"].isna().sum())

print("\nSource resource zero values:")
print("requested CPU:", (df["requested_cpu_ncu"] == 0).sum())
print("requested memory:", (df["requested_memory_normalized"] == 0).sum())
print("peak CPU:", (df["peak_cpu_ncu"] == 0).sum())
print("peak memory:", (df["peak_memory_normalized"] == 0).sum())

print("\nExecution-level zero peaks:")
print("peak CPU:", (execution_df["peak_cpu_ncu"] == 0).sum())
print(
    "peak memory:",
    (execution_df["peak_memory_normalized"] == 0).sum()
)

# Check whether any execution/time windows contain a mixture
# of known and missing measurements.
window_missing_check = (
    df.groupby(
        [
            "execution_id",
            "usage_start_time_us",
            "usage_end_time_us",
        ]
    )
    .agg(
        cpu_count=("peak_cpu_ncu", "size"),
        cpu_known=("peak_cpu_ncu", "count"),
        memory_count=("peak_memory_normalized", "size"),
        memory_known=("peak_memory_normalized", "count"),
    )
)

partial_cpu_windows = (
    (window_missing_check["cpu_known"] > 0)
    & (window_missing_check["cpu_known"] < window_missing_check["cpu_count"])
)

partial_memory_windows = (
    (window_missing_check["memory_known"] > 0)
    & (
        window_missing_check["memory_known"]
        < window_missing_check["memory_count"]
    )
)

print("\nPartially missing time windows:")
print("CPU:", partial_cpu_windows.sum())
print("Memory:", partial_memory_windows.sum())

Source resource missing values:
requested CPU: 774
requested memory: 774
peak CPU: 0
peak memory: 4927

Source resource zero values:
requested CPU: 6287
requested memory: 202
peak CPU: 46176
peak memory: 8649

Execution-level zero peaks:
peak CPU: 13868
peak memory: 774

Partially missing time windows:
CPU: 0
Memory: 0


## Step 10 — Validate Execution-Level Records

This validation confirms that every reconstructed execution has a unique ID, a valid runtime, and valid observation and machine counts.

In [11]:
# Validate execution IDs.
print(
    "Unique execution IDs:",
    execution_df["execution_id"].nunique()
)

print(
    "Duplicate execution IDs:",
    execution_df["execution_id"].duplicated().sum()
)

# Validate runtime values.
print(
    "Negative runtimes:",
    (execution_df["runtime_us"] < 0).sum()
)

print(
    "Missing runtimes:",
    execution_df["runtime_sec"].isna().sum()
)

# Summarize execution runtimes.
print("\nRuntime summary in seconds:")
display(
    execution_df["runtime_sec"].describe()
)

# Summarize the number of observations per execution.
print("\nObservations per execution:")
display(
    execution_df["observation_count"].describe()
)

# Summarize the number of machines per execution.
print("\nMachines per execution:")
display(
    execution_df["machine_count"].describe()
)

Unique execution IDs: 253585
Duplicate execution IDs: 0
Negative runtimes: 0
Missing runtimes: 0

Runtime summary in seconds:


count      253585.0
mean     225.343475
std      122.660864
min             1.0
25%           105.0
50%           300.0
75%           300.0
max           300.0
Name: runtime_sec, dtype: Float64


Observations per execution:


count    253585.000000
mean          1.600623
std          15.606338
min           1.000000
25%           1.000000
50%           1.000000
75%           1.000000
max        6566.000000
Name: observation_count, dtype: float64


Machines per execution:


count    253585.000000
mean          1.600595
std          15.606330
min           1.000000
25%           1.000000
50%           1.000000
75%           1.000000
max        6566.000000
Name: machine_count, dtype: float64

## Step 11 — Attach Workload Identity

Each reconstructed execution is linked to a recurring workload using:

- `user`
- `collection_logical_name`
- `collection_name` as a fallback

The resulting `recurrence_key` identifies the recurring workload family. Multiple execution IDs may belong to the same recurrence key.

In [12]:
# -----------------------------------------
# Step 11: Attach workload identity
# -----------------------------------------

# Remove identity columns if this cell is rerun
execution_df = execution_df.drop(
    columns=[
        "user",
        "collection_name",
        "collection_logical_name",
        "logical_workload_name",
        "user_clean",
        "recurrence_key",
        "recurring_job_id",
    ],
    errors="ignore",
)

# Aggregate identity information at execution level
identity_columns = [
    column
    for column in [
        "user",
        "collection_name",
        "collection_logical_name",
    ]
    if column in df.columns
]

if identity_columns:
    execution_identity = (
        df.groupby("execution_id", as_index=False)[identity_columns]
        .first()
    )
else:
    execution_identity = execution_df[["execution_id"]].copy()

# Attach identity columns
execution_df = execution_df.merge(
    execution_identity,
    on="execution_id",
    how="left",
    validate="one_to_one",
)

# Clean user identity
if "user" in execution_df.columns:
    execution_df["user_clean"] = (
        execution_df["user"]
        .astype("string")
        .fillna("unknown_user")
    )
else:
    execution_df["user_clean"] = "unknown_user"

# Prefer logical collection name
if "collection_logical_name" in execution_df.columns:
    execution_df["logical_workload_name"] = (
        execution_df["collection_logical_name"]
        .astype("string")
    )

    if "collection_name" in execution_df.columns:
        execution_df["logical_workload_name"] = (
            execution_df["logical_workload_name"]
            .fillna(
                execution_df["collection_name"].astype("string")
            )
        )

elif "collection_name" in execution_df.columns:
    execution_df["logical_workload_name"] = (
        execution_df["collection_name"]
        .astype("string")
    )

else:
    execution_df["logical_workload_name"] = "unknown_job"

execution_df["logical_workload_name"] = (
    execution_df["logical_workload_name"]
    .fillna("unknown_job")
)

# Construct recurring workload key
execution_df["recurrence_key"] = (
    execution_df["user_clean"].astype("string")
    + "||"
    + execution_df["logical_workload_name"].astype("string")
)

# Alias used in later notebooks
execution_df["recurring_job_id"] = execution_df["recurrence_key"]

# -----------------------------------------
# Validate final execution table
# -----------------------------------------

print("Final execution records:", len(execution_df))
print(
    "Unique execution IDs:",
    execution_df["execution_id"].nunique()
)
print(
    "Missing execution IDs:",
    execution_df["execution_id"].isna().sum()
)
print(
    "Missing recurrence keys:",
    execution_df["recurrence_key"].isna().sum()
)
print(
    "Unique recurrence keys:",
    execution_df["recurrence_key"].nunique()
)

print("\nExecutions per recurrence key:")
display(
    execution_df
    .groupby("recurrence_key")
    .size()
    .describe()
)

# Preview final columns
display(
    execution_df[
        [
            "execution_id",
            "collection_id",
            "instance_index",
            "candidate_instance_key",
            "episode_number",
            "recurrence_key",
            "execution_start_time_us",
            "execution_end_time_us",
            "runtime_sec",
            "observation_count",
            "machine_count",
        ]
    ].head()
)

Final execution records: 253585
Unique execution IDs: 253585
Missing execution IDs: 0
Missing recurrence keys: 0
Unique recurrence keys: 2280

Executions per recurrence key:


count     2280.000000
mean       111.221491
std        725.639610
min          1.000000
25%          1.000000
50%          4.000000
75%         28.000000
max      25180.000000
dtype: float64

,execution_id,collection_id,instance_index,candidate_instance_key,episode_number,recurrence_key,execution_start_time_us,execution_end_time_us,runtime_sec,observation_count,machine_count
0,100286199||134||episode_1,100286199,134,100286199||134,1,iTVkGxgmnzN6bSosM7QLqyRpmOAxn1/H+cA2UBz/20g=||...,2292000000000,2292300000000,300.0,1,1
1,100286199||250||episode_1,100286199,250,100286199||250,1,iTVkGxgmnzN6bSosM7QLqyRpmOAxn1/H+cA2UBz/20g=||...,2292000000000,2292300000000,300.0,1,1
2,100286199||269||episode_1,100286199,269,100286199||269,1,iTVkGxgmnzN6bSosM7QLqyRpmOAxn1/H+cA2UBz/20g=||...,2292000000000,2292300000000,300.0,1,1
3,100286199||321||episode_1,100286199,321,100286199||321,1,iTVkGxgmnzN6bSosM7QLqyRpmOAxn1/H+cA2UBz/20g=||...,2292000000000,2292300000000,300.0,1,1
4,100286199||323||episode_1,100286199,323,100286199||323,1,iTVkGxgmnzN6bSosM7QLqyRpmOAxn1/H+cA2UBz/20g=||...,2292000000000,2292300000000,300.0,1,1


## Step 12 — Save Reconstructed Executions

The finalized execution-level dataset is saved as the canonical Google subset execution artifact for downstream pipeline stages.

The output contains one row per reconstructed execution, including execution metadata, aggregated CPU and memory resources, runtime metrics, and recurring workload identity.

In [13]:
# -----------------------------------------
# Step 12: Save reconstructed executions
# -----------------------------------------

EXECUTIONS_PATH = (
    REPO_ROOT
    / "data"
    / "processed"
    / "google_subset_v1"
    / "google_subset_executions_v1.parquet"
)

EXECUTIONS_PATH.parent.mkdir(parents=True, exist_ok=True)

execution_df.to_parquet(
    EXECUTIONS_PATH,
    index=False
)

print("Saved reconstructed execution dataset:")
print(EXECUTIONS_PATH)
print("Shape:", execution_df.shape)

Saved reconstructed execution dataset:
c:\Users\012320802\masters-project\data\processed\google_subset_v1\google_subset_executions_v1.parquet
Shape: (253585, 26)


## Step 13 — Validate Saved Execution Artifact

The saved execution artifact is reloaded and validated to confirm that the final output preserves the expected execution structure and resource aggregation.

Validation checks include:

- unique execution IDs
- missing execution and recurrence identifiers
- negative runtimes
- reconstructed episode counts
- candidate groups with multiple episodes
- missing requested and peak CPU/memory values

In [14]:
# -----------------------------------------
# Step 13: Validate saved execution artifact
# -----------------------------------------

# Reload the saved execution-level dataset
saved_execution_df = pd.read_parquet(EXECUTIONS_PATH)

print("Saved dataset shape:", saved_execution_df.shape)
print("Unique execution IDs:",
      saved_execution_df["execution_id"].nunique())
print("Duplicate execution IDs:",
      saved_execution_df["execution_id"].duplicated().sum())
print("Missing execution IDs:",
      saved_execution_df["execution_id"].isna().sum())
print("Missing recurrence keys:",
      saved_execution_df["recurrence_key"].isna().sum())
print("Negative runtimes:",
      (saved_execution_df["runtime_us"] < 0).sum())

print("\nExecution records by episode number:")

episode_summary = (
    saved_execution_df
    .groupby("episode_number")
    .size()
    .reset_index(name="execution_count")
)

display(episode_summary)

print("\nCandidate groups with multiple episodes:")

multiple_episode_groups = (
    saved_execution_df
    .groupby("candidate_instance_key")["episode_number"]
    .nunique()
    .reset_index(name="episode_count")
)

multiple_episode_groups = multiple_episode_groups[
    multiple_episode_groups["episode_count"] > 1
]

print("Count:", len(multiple_episode_groups))

display(
    multiple_episode_groups
    .sort_values("episode_count", ascending=False)
    .head(10)
)

print("\nSaved columns:")
print(saved_execution_df.columns.tolist())


print("\nSaved resource validation:")
print(
    "Missing requested CPU:",
    saved_execution_df["requested_cpu_ncu"].isna().sum()
)
print(
    "Missing requested memory:",
    saved_execution_df["requested_memory_normalized"].isna().sum()
)
print(
    "Missing peak CPU:",
    saved_execution_df["peak_cpu_ncu"].isna().sum()
)
print(
    "Missing peak memory:",
    saved_execution_df["peak_memory_normalized"].isna().sum()
)

Saved dataset shape: (253585, 26)
Unique execution IDs: 253585
Duplicate execution IDs: 0
Missing execution IDs: 0
Missing recurrence keys: 0
Negative runtimes: 0

Execution records by episode number:


,episode_number,execution_count
0,1,242946
1,2,10291
2,3,340
3,4,8



Candidate groups with multiple episodes:
Count: 10291


,candidate_instance_key,episode_count
49175,260697606809||667,4
30899,221495509609||651,4
48950,260697606809||1114,4
49177,260697606809||676,4
49187,260697606809||749,4
49060,260697606809||2032,4
49208,260697606809||933,4
49018,260697606809||1633,4
30925,221495509609||901,3
30778,221495509609||1875,3



Saved columns:
['execution_id', 'collection_id', 'instance_index', 'candidate_instance_key', 'episode_number', 'execution_start_time_us', 'execution_end_time_us', 'observation_count', 'machine_count', 'average_cpu_ncu', 'average_memory_normalized', 'requested_cpu_ncu', 'requested_memory_normalized', 'peak_cpu_ncu', 'peak_memory_normalized', 'runtime_us', 'runtime_sec', 'slack_cpu_ncu', 'slack_memory_normalized', 'user', 'collection_name', 'collection_logical_name', 'user_clean', 'logical_workload_name', 'recurrence_key', 'recurring_job_id']

Saved resource validation:
Missing requested CPU: 774
Missing requested memory: 774
Missing peak CPU: 0
Missing peak memory: 4495


## Step 14 — Audit Recurring Workload Groups

A `recurrence_key` groups execution episodes belonging to the same logical workload. This cell checks how many executions belong to each recurring workload.

In [15]:
# Number of executions per recurring workload
recurrence_summary = (
    saved_execution_df
    .groupby("recurrence_key")
    .agg(
        execution_count=("execution_id", "nunique"),
        first_execution_time_us=("execution_start_time_us", "min"),
        last_execution_time_us=("execution_end_time_us", "max"),
        candidate_group_count=("candidate_instance_key", "nunique"),
    )
    .reset_index()
)

print("Unique recurrence keys:",
      recurrence_summary["recurrence_key"].nunique())

print("Recurrence keys with multiple executions:",
      (recurrence_summary["execution_count"] > 1).sum())

print("\nExecutions per recurrence key:")
display(
    recurrence_summary["execution_count"].describe()
)

print("\nMost frequent recurring workloads:")
display(
    recurrence_summary
    .sort_values("execution_count", ascending=False)
    .head(20)
)

Unique recurrence keys: 2280
Recurrence keys with multiple executions: 1504

Executions per recurrence key:


count     2280.000000
mean       111.221491
std        725.639610
min          1.000000
25%          1.000000
50%          4.000000
75%         28.000000
max      25180.000000
Name: execution_count, dtype: float64


Most frequent recurring workloads:


,recurrence_key,execution_count,first_execution_time_us,last_execution_time_us,candidate_group_count
794,DrrEIEWkWuW7RrZwpHLCN0k0A2J0usJeyt3wtqzZ7Kk=||...,25180,600000000,2677945000000,23783
1861,ly4HE4d+6E8BqXgm6E1eGNKu8OiiZE+Gj/D4YUMRFDU=||...,11457,46500000000,2655600000000,10913
777,DHJYL1IYxKn7+hgqicbUBV4/cAvCGBBmmpO+F9Rb1P4=||...,7893,194700000000,2124300000000,6371
1913,ly4HE4d+6E8BqXgm6E1eGNKu8OiiZE+Gj/D4YUMRFDU=||...,7745,41613000000,2644500000000,6188
1078,N4z3CQhsZb0Z/6KYX+6Y6CyETgv5iobfcL+/h+RZeA4=||...,7098,305700000000,2444700000000,7098
1647,fyVrWxhkLIMjl2J3LD3cpOWRuMUdb9sqYAI1OSXrPX0=||...,7071,952500000000,2521200000000,7071
795,DrrEIEWkWuW7RrZwpHLCN0k0A2J0usJeyt3wtqzZ7Kk=||...,5862,23100000000,2660100000000,5125
1783,jW2GJpOlCheoSzGcwfmRjkjV6mYZo4DC43ej/X/OyjM=||...,4928,1200000000,2663400000000,4612
13,+OVSEukSmo6ma+28nY4raBvhviWrpQLsc+DWQcTGV4E=||...,3323,2354603000000,2354700000000,3323
1520,bIEfKw4txb2bIvwdWUnKf0wrWg3TY2zGk85oI4/bK8k=||...,3093,1504505000000,1504800000000,3093


## Logical Workload Identity

A logical workload family represents a recurring job or workload that may run multiple times.

It is identified using:

~~~text
recurrence_key = user_clean || logical_workload_name
~~~

The `||` symbol separates the user from the logical workload name.

### Audit Results

- **2,280 recurrence keys** were identified.
- These represent **2,280 inferred logical workload families**.
- **1,504 recurrence keys** had multiple executions.
- **776 recurrence keys** had only one execution.
- Each recurrence key had an average of **111.2 executions**.
- The most frequent workload had **25,180 execution episodes**.

### Key Definitions

| Concept | Definition | Key |
|---|---|---|
| Raw observation | One machine-level usage record from the trace | `observation_id` |
| Candidate group | Initial group of potentially related observations | `collection_id || instance_index` |
| Execution episode | One continuous activity period within a candidate group | `candidate_instance_key || episode_number` |
| Execution ID | Unique identifier for one reconstructed execution | `execution_id` |
| Logical workload family | Group of repeated executions belonging to the same user and logical job | `user_clean || logical_workload_name` |
| Recurrence key | Identifier used to group repeated executions | `recurrence_key` |

### Candidate Groups and Execution Episodes

A candidate group is created using:

~~~text
candidate_instance_key = collection_id || instance_index
~~~

A candidate group may contain one or more separated activity periods. When a sufficiently large time gap exists between observations, the group is divided into separate episodes.

For example:

~~~text
One candidate group
        |
        |-- episode_1
        |-- episode_2
        |-- episode_3
~~~

Each episode receives its own execution ID:

~~~text
execution_id = candidate_instance_key || episode_number
~~~

### Example

The most frequent logical workload family had:

- **25,180 reconstructed execution episodes**
- **23,783 candidate groups**

The number of executions is greater than the number of candidate groups because some candidate groups contained multiple separated episodes.

Therefore:

~~~text
23,783 candidate groups
        →
25,180 reconstructed execution episodes
~~~

This does **not** mean that there were 25,180 different workload types. It means that one logical workload family ran 25,180 separate times or activity episodes.

### Relationship Between the Identifiers

~~~text
Raw observations
        ↓
collection_id || instance_index
        ↓
Candidate group
        ↓
Time-gap based episode reconstruction
        ↓
episode_1, episode_2, episode_3, ...
        ↓
Unique execution_id
        ↓
user_clean || logical_workload_name
        ↓
recurrence_key
        ↓
Logical workload family
~~~

### Important Distinction

The `user` field was used only to construct the `recurrence_key`. It was not combined with the `execution_id`.

Therefore:

- `execution_id` identifies one reconstructed execution.
- `recurrence_key` groups repeated executions of the same logical workload.
- Multiple execution IDs may belong to one recurrence key.
- Different users with similarly named jobs remain separate workload families.
- The recurrence key represents an inferred workload grouping, not an explicit ground-truth job ID from the dataset.